### ISSUE TITLE
wyze bulb color

### ISSUE THREAD

**Issue by dgmltn** [noelportugal/wyze-node#15]
Will this work with Wyze bulb color? `setProperty(mac, model, "P1507", "0000ff")` for instance doesn't seem to affect my bulb, or return predictably.

**Comment by noelportugal** [comment]
I haven't received my bulb color yet, but will test as soon as I get em!

**Comment by nosaj72** [comment]
I've been playing with this today, and I'm not able to get any of the **setProperty** calls to seem to do anything to the color bulb. **turnOn** and **turnOff** work, however.

**Comment by nathanctech** [comment]
After MITMing the requests from the app itself, I've written this:

```js
  async runActionList(instance, provider, actionKey, actionParamsList) {
    let result
    try {
      await this.getTokens();
      if (!this.accessToken) {
        await this.login()
      }

      const data = {
        action_list:[
          {
            action_key: actionKey,
            action_params: { list: actionParamsList },
            instance_id: instance,
            provider_key: provider
          }],
          
      }

      result = await axios.post(`${this.baseUrl}/app/v2/auto/run_action_list`, await this.getRequestBodyData(data))

      if (result.data.msg === 'AccessTokenError') {
        await this.getRefreshToken()
        return this.runActionList(instance, provider, actionKey, actionParamsList)
      }
    }
    catch (e) {
      throw e
    }
    return result.data
  }
  ```
  
  It appears the color bulbs use this action list parameter instead, like so:
  
  ```
      "action_list": [
        {
            "action_key": "set_mesh_property",
            "action_params": {
                "list": [
                    {
                        "mac": "<mac>",
                        "plist": [
                            {
                                "pid": "P1507",
                                "pvalue": "00ff00"
                            }
                        ]
                    }
                ]
            },
            "instance_id": "<mac>",
            "provider_key": "WLPA19C"
        }
        ```
Weirdly, I got it to respond exactly once, then the bulbs stopped responding.

For what it's worth, if you're on the same network as the bulbs, you can POST to `http://<local ip>/device_request with some simple json to get a color change (or turn it off/on):

```js
{
    "characteristics": [
        {
            "pid": "P1507",
            "pvalue": "ffe207"
        }
    ],
    "isSendQueue": 0,
    "request": "set_status"
}
```

That's about as far as I've gotten. Need to mess with the bulbs a bit more, for sure!

**Comment by dgmltn** [comment]
I noticed the local api too, and these commands seem to work well... until you start using the wyze app to control the same bulb. At that point, the bulb stops responding to these requests from this device for a while.

```
# on
curl -vv -X POST -H 'root-response:true' -H 'Mesh-Node-Mac:[mac]' -H 'Connection:close' -H 'Mesh-Node-Num:1' -H 'Content-Type:application/json' -H 'Host:[ip address]' -H 'User-Agent:okhttp/4.7.2' -d '{"request":"set_status","isSendQueue":0,"characteristics":[{"pid":"P3","pvalue":"1"}]}' 'http://[ip address]/device_request'
```

```
# green
curl -X POST -H "root-response:true" -H "Mesh-Node-Mac:[mac]" -H "Connection:close" -H "Mesh-Node-Num:1" -H "Content-Type:application/json" -H "Host:[ip address]" -H "Accept-Encoding:gzip" -H "User-Agent:okhttp/4.7.2" -d '{"request":"set_status","isSendQueue":0,"characteristics":[{"pid":"P1507","pvalue":"00ff00"}]}' "http://[ip address]/device_request"
```
```
# white
curl -X POST -H "root-response:true" -H "Mesh-Node-Mac:[mac]" -H "Connection:close" -H "Mesh-Node-Num:1" -H "Content-Type:application/json" -H "Host:[ip address]" -H "User-Agent:okhttp/4.7.2" -d '{"request":"set_status","isSendQueue":0,"characteristics":[{"pid":"P1502","pvalue":"6263"}]}' "http://[ip address]/device_request"

```

**Comment by shauntarves** [comment]
@dgmltn if you are interested in color bulb and other devices, please have a look at my recently released [Python SDK for Wyze](https://github.com/shauntarves/wyze-sdk). It has support for a variety of devices and device features.

**Comment by burnhamd** [comment]
> I noticed the local api too, and these commands seem to work well... until you start using the wyze app to control the same bulb. At that point, the bulb stops responding to these requests from this device for a while.
> 
> ```
> # on
> curl -vv -X POST -H 'root-response:true' -H 'Mesh-Node-Mac:[mac]' -H 'Connection:close' -H 'Mesh-Node-Num:1' -H 'Content-Type:application/json' -H 'Host:[ip address]' -H 'User-Agent:okhttp/4.7.2' -d '{"request":"set_status","isSendQueue":0,"characteristics":[{"pid":"P3","pvalue":"1"}]}' 'http://[ip address]/device_request'
> ```
> 
>  ```
> # green
> curl -X POST -H "root-response:true" -H "Mesh-Node-Mac:[mac]" -H "Connection:close" -H "Mesh-Node-Num:1" -H "Content-Type:application/json" -H "Host:[ip address]" -H "Accept-Encoding:gzip" -H "User-Agent:okhttp/4.7.2" -d '{"request":"set_status","isSendQueue":0,"characteristics":[{"pid":"P1507","pvalue":"00ff00"}]}' "http://[ip address]/device_request"
> ```
> 
>  ```
> # white
> curl -X POST -H "root-response:true" -H "Mesh-Node-Mac:[mac]" -H "Connection:close" -H "Mesh-Node-Num:1" -H "Content-Type:application/json" -H "Host:[ip address]" -H "User-Agent:okhttp/4.7.2" -d '{"request":"set_status","isSendQueue":0,"characteristics":[{"pid":"P1502","pvalue":"6263"}]}' "http://[ip address]/device_request"
> ```
> 
> Edit: It looks like the bulb doesn't even care about any of those headers. Just this request also works fine:
> 
> ```
> # on
> curl -X POST -d '{"request":"set_status","isSendQueue":0,"characteristics":[{"pid":"P3","pvalue":"1"}]}' 'http://[ip address]/device_request'
> ```

Local API is interesting.  I am getting connection refused for any call I make though.  Is there a change in newer firmware?

**Comment by dgmltn** [comment]
> Local API is interesting. I am getting connection refused for any call I make though. Is there a change in newer firmware?

Yeah, my system stopped working with the latest firmware update 😠

**Comment by nathanctech** [comment]
Yes, the port changed to 8081 and now the payload is encrypted.

**Comment by dgmltn** [comment]
> Yes, the port changed to 8081 and now the payload is encrypted.

Did you figure out how to fix that yet? Or maybe how to downgrade the bulbs?

**Comment by burnhamd** [comment]
> Yes, the port changed to 8081 and now the payload is encrypted.

So just make a https request with same payload or are you saying the payload itself has a separate encryption?
